cpp/ParseFasm.cpp

```diff
@@ -0,0 +1,414 @@
+    long long integer = 0;
+    std::string value = context->DECIMAL_VALUE()->getText();
+    auto it = value.begin();
+    it += 2; // skip 'd
+
+    // Build up a Num, skipping '_'.
+    while (it != value.end()) {
+      if (*it != '_') {
+        integer = (integer * 10) + (*it - '0');
+      }
+      it++;
+    }
+
+    std::ostringstream data;
+    data << Num('d', integer);
+    return data.str();
+  }
+
+  // A Verilog octal value.
+  // Tags: octal (o)
+  virtual Any visitOctalValue(FasmParser::OctalValueContext *context) override {
+    std::ostringstream data;
+    std::string value = context->OCTAL_VALUE()->getText();
+    auto it = value.begin();
+    it += 2; // skip 'b
+
+    // Build up a Num 3 bits at a time.
+    // Note that since the word size is not evenly divisible by 3,
+    // intermediate values can be greater than the word size.
+    // This is why the 'word' below is 64 bits wide.
+    int bits = lead_bits(count_without(it, value.end(), '_') * 3);
+    uint64_t word = 0; // could temporarily overflow uint32_t
+    while (it != value.end()) {
+      if (*it != '_') {
+        word = (word << 3) | (*it - '0');
+        bits += 3;
+        if (bits >= Num::width) {
+          data << Num(word >> (bits - Num::width));
+          word >>= Num::width;
+          bits -= Num::width;
+        }
+      }
+      it++;
+    }
+    assert(!word);
+    return withHeader('o', data.str());
+  }
+
+  // A collection of annotations. { ... }
+  // Tags: annotations ({)
+  virtual Any visitAnnotations(FasmParser::AnnotationsContext *context) override {
+    std::ostringstream data;
+    for (auto &a : context->annotation()) {
+      data << visit(a).as<std::string>();
+    }
+    return withHeader('{', data.str());
+  }
+
+  // An annotation: x = "y"
+  // Tags: annotation (a), annotation name (.), annotation value (=)
+  virtual Any visitAnnotation(FasmParser::AnnotationContext *context) override {
+    std::ostringstream data;
+    data << Str('.', context->ANNOTATION_NAME()->getText());
+    if (context->ANNOTATION_VALUE()) {
+      std::string value = context->ANNOTATION_VALUE()->getText();
+      value.erase(0, 1);  // Convert "value" -> value
+      value.pop_back();
+      data << Str('=', value);
+    }
+    return withHeader('a', data.str());
+  }
+
+private:
+
+  std::ostream &out;
+};
+
+
+// Common portion of 'from_string' and 'from_file'.
+// Consumes an input stream and produces an output stream.
+static
+void parse_fasm(std::istream &in, std::ostream &out) {
+  ANTLRInputStream stream(in);
+  FasmLexer lexer(&stream);
+  CommonTokenStream tokens(&lexer);
+  FasmParser parser(&tokens);
+  auto *tree = parser.fasmFile();
+  FasmParserBaseVisitor(out).visit(tree);
+}
+
+// This static string is used to temporarily hold the result
+// returned from external functions.
+// This is to allow Python consumers, because I could not
+// find a convenient way for Python to free a malloc'ed string.
+static std::string result;
```

Comment:
Please fix this, it is pretty straight forward: See https://stackoverflow.com/questions/13445568/python-ctypes-how-to-free-memory-getting-invalid-pointer-error

Basically add a third method to your interface to free the memory.

Reply:
It means I either I need to copy the `std::string` to a malloc'ed `char *`, or add two methods: one to free a `std::string`, (returned as a `void *` from the other methds), and one to get a `char *` from that e.g.:

```python
p = from_string(...)
data = from_std_string(p)
...
free(p)
```

Reply:
Just use `strdup` and `free`?  Given that there is only one copy per file, I don't imagine the runtime or memory costs will be too bad.  I'm also fine with the the `get char* from std::string *` and `delete std::string *`.

Reply:
I came up with a better solution: Python passes a callback into `from_*` and runs before the string buffer is freed, so there is no need for manual memory management in Python or in C++.

The callback holds a reference that it uses to store the resulting model.